feat(tables): migrate react-tables to TS

austin94 opened the pull request:
## Description

This PR migrates `react-tables` to TS. It includes some heavy-rewrites of the examples to support the default `<table>` styling that is now used.

## Breaking changes

```
## @zendeskgarden/react-tables

- no longer packages a `styles.css` dist; CSS is self-contained
- All components now use standard `<table>` styling
- Removed all `focused`, `active`, and `hovered` states from components
- Component props renames:
  - `Cell`
    - `minimum` -> `isMinimum`
    - `truncate` -> `isTruncated`
    - `menu` -> `hasOverflow`
  - `Row`
    - `striped` -> `isStriped`
    - `selected` -> `isSelected`
```

## Checklist

- [ ] :ok_hand: design updates are Garden Designer approved (add the
      designer as a reviewer)
- [x] :nail_care: view component styling is based on a Garden CSS
      component
- [x] :globe_with_meridians: Styleguidist demo is up-to-date (`yarn start`)
- [x] :arrow_left: renders as expected with reversed (RTL) direction
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :guardsman: includes new unit tests
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

hzhu commented on packages/tables/examples/selectable-table.md:
```diff
@@ -66,7 +66,7 @@ const isSelectAllChecked = (selectedRows, rows) => {
   <Body>
     {state.rows.map(row => (
       <Row key={row.id} selected={state.selectedRows[row.id]}>
```
I noticed that the selected row styling was missing. I think renaming the `selected` prop to `isSelected` would fix it.

Notice: styling is not applied:
![TruncatedTitle](https://user-images.githubusercontent.com/1811365/70356963-82c71680-182a-11ea-9457-30118fd826bb.gif)

hzhu commented on packages/tables/examples/paginated-table.md:
```diff
@@ -36,34 +36,30 @@ const getPagedData = (data, currentPage, pageSize) => {
-      <XL tag={Caption} style={{ marginBottom: DEFAULT_THEME.space.sm }}>
-        Paginated Tickets
-      </XL>
+    <Caption style={{ marginBottom: DEFAULT_THEME.space.sm }}>
```
I noticed that in production, the margin bottom is applied to the `<XL>` immediately below this line. What do you think about removing the margin here, and applying it back to the `<XL>`?

Before (production): 
<img width="1162" alt="Screen Shot 2019-12-06 at 2 36 42 PM" src="https://user-images.githubusercontent.com/1811365/70361423-6630db80-1836-11ea-9a3b-c4f58f983c5c.png">

Current (feature branch), the spacing between the `<Caption>` and `<XL>` is very close:
<img width="1122" alt="Screen Shot 2019-12-06 at 2 40 34 PM" src="https://user-images.githubusercontent.com/1811365/70361424-6630db80-1836-11ea-8484-88f3994b2e0e.png">

jzempel commented:
@austin94 please include @allisonacs for example review coverage and please publish the demo to ease the review for all of us. Thanks.

hzhu commented on packages/tables/styleguide.config.js:
```diff
@@ -5,61 +5,69 @@
     {
-      name: 'Tables',
-      content: '../../packages/tables/src/examples/tables.md',
+      name: 'Examples',
```
This is an open ended question. I noticed that early on, CSS-in-JS example sections are organized like this:

```
Examples
├── Basic
└── Advanced
Elements
├── Element 1
├── Element 2
└── Element 3
```

Should the examples for `react-tables` be organized under "Basic" and "Advanced" too? Or is `react-tables` complex enough to require a different organization strategy? When would one know _not_ to follow the above organization (e.g. organizing under a "Basic", "Advanced", etc)?

austin94 replied:
I've included the basic styling/format props as knobs in the first basic example. All of the other examples include a lot of custom logic and styling for their use-case.

austin94 replied:
Additionally, having the separated example naming helps with readability. If I remember correctly the "Virtual scrolling" example is the most shared deep-link across all the documentation.

hzhu replied:
Thanks for the clarification!

austin94 commented:
Pre-published at https://austin-next.netlify.com/tables/

jzempel commented on packages/tables/src/elements/OverflowButton.tsx:
```diff
@@ -0,0 +1,43 @@
+OverflowButton.propTypes = {
+  isHovered: PropTypes.bool,
+  isActive: PropTypes.bool,
+  isFocused: PropTypes.bool
```
Do we see this getting cleaned-up with CSS-in-JS?

austin94 replied:
I was planning to address this in the CSS-in-JS PR to ensure I bring over all the necessary styling.

jzempel commented on packages/tables/src/elements/Row.tsx:
```diff
@@ -0,0 +1,43 @@
+Row.propTypes = {
+  isStriped: PropTypes.bool,
+  isFocused: PropTypes.bool,
+  isHovered: PropTypes.bool,
```
Why do we want to expose these props at the element level? We don't want folks manually tweaking hover/focus states.

austin94 replied:
Overall, I was planning on addressing these API removals in the CSS-in-JS PR.

But, `Row` may need to be our exception here. There are several use-cases where a consumer may need to manually apply this styling based on a table state. The `drag-and-drop` example is a good use-case for making the `isHovered` prop available.

jzempel commented on packages/tables/src/elements/SortableCell.tsx:
```diff
@@ -0,0 +1,44 @@
+SortableCell.propTypes = {
+  sort: PropTypes.oneOf(['asc', 'desc']),
+  isFocused: PropTypes.bool,
+  isActive: PropTypes.bool,
```
Same comment here re: not wanting users to manually control focus/active states.

austin94 replied:
I was planning to address this in the CSS-in-JS PR to ensure I bring over all the necessary styling.

jzempel commented on packages/tables/src/elements/Table.tsx:
```diff
@@ -0,0 +1,23 @@
+>((props, ref) => <StyledTable ref={ref} {...props} />);
+
+Table.propTypes = {
+  size: PropTypes.oneOf(['small', 'large'])
```
Following on other components, we should have a `medium` (default).

jzempel commented on packages/tables/src/styled/StyledHeaderCell.ts:
```diff
@@ -0,0 +1,19 @@
+
+const COMPONENT_ID = 'tables.header_cell';
+
+export const StyledHeaderCell = styled(StyledCell.withComponent('th')).attrs({
```
Prefer `as` attr vs the `withComponent` deprecation candidate.

austin94 replied:
I've moved to the `as` prop in `attrs`.

Looking through the docs, [forwardedAs](https://www.styled-components.com/docs/api#forwardedas-prop) also looks really useful. We will have to watch our `peerDependency` range for some of these utilities as we are on the low-end for some of these props.

austin94 commented on packages/tables/examples/basic.md:
```diff
@@ -0,0 +1,162 @@
+  <StyledSpacer />
+  <Layout.Row>
+    <Layout.Col>
+      <Table size={state.rowSize === 'default' ? undefined : state.rowSize}>
```
With the examples listed about I ended up with a large if statement which rendered 5 unique examples. I did several iterations and ended up combining some of the simpler examples, but some of these needed to be split out to make them usable as base-line examples.

jzempel replied:
I still feel like there's a both/and option here. Do the combo thing, but break out the individual examples for code readability. This is the direction I would be trying to take this example, but YMMV.

hzhu commented on packages/tables/src/elements/Table.tsx:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Accepts all `<table>` attributes and events
+ */
+export const Table = React.forwardRef<
```
I noticed that there are [existing](https://github.com/zendeskgarden/react-components/blob/next/packages/chrome/src/elements/Chrome.spec.tsx#L26) [unit](https://github.com/zendeskgarden/react-components/blob/next/packages/grid/src/elements/Grid.spec.tsx#L19) [tests](https://github.com/zendeskgarden/react-components/blob/next/packages/tabs/src/elements/Tab.spec.tsx#L18) which assert that components exported to consumers should forward refs.

Should this PR include unit test for `Table` ( and `Body`, `Caption`, `Cell`, `GroupRow`, `Head`, `HeaderCell`, `HeaderRow`, `OverflowButton`, `Row`) now that these components have been updated to support ref forwarding?

austin94 replied:
I've now added these `ref` checks. I also found a invalid DOM nesting warnings from `react-dom` (thanks react-testing-library 🎉 ). I've updated the `Cell/Row` examples to include wrapping `Table/Body` elements

austin94 commented:
> I would expect the blue color to change back to grey if I move my mouse off the control but it stays blue bc that's where the browser focus is. Can we fix that?

We can address this in the CSS-in-JS PR if we feel it is necessary.

> Maybe a toggle to turn the caption on/off, and something that inserts/removes the <MD> as well?

I've added a toggle to the example and modified the styling to use our theming values.

allisonacs commented:
> > I would expect the blue color to change back to grey if I move my mouse off the control but it stays blue bc that's where the browser focus is. Can we fix that?
> 
> We can address this in the CSS-in-JS PR if we feel it is necessary.

Yea, let's make sure we do this. It feels off right now.

allisonacs commented:
@austin94 the first part of my comment was that it seemed like there were more individual examples than I would have anticipated. Is there a way to cleanly cut down on the number of examples?

austin94 commented:
> Is there a way to cleanly cut down on the number of examples?

I don't see a way to combine the remaining example while still allowing developers to use them as an implementation reference. Each of them require custom rendering and styling logic that is complex enough to not be shared.

If we want to spend more time combining these examples I believe it would be best to estimate it in a separate PR.
